### PR TITLE
test: Add integration tests 

### PR DESCRIPTION
**Pull request by dheerajsngh**
* Add integration tests for the Analytics core integration tests

### CODE REVIEW

File: gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopOutputStreamAnalyticsIntegrationTest.java [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+    GoogleHadoopOutputStream out = createGhfsOutputStream(ghfs, path, CreateFileOptions.DEFAULT);
+    out.close();
+
+    assertThrows(IOException.class, () -> out.write(1));
+  }
+
+  @Test
+  public void write_zeroBytes_createsEmptyFile() throws Exception {
+    URI path = gcsFsIHelper.getUniqueObjectUri(getClass(), "write_zeroBytes");
+    GoogleHadoopFileSystem ghfs = setupGhfs("write_zeroBytes");
```
[the remark this answers is not in the document]
Comment (dheerajsngh) [reply gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopOutputStreamAnalyticsIntegrationTest.java]: Done.

File: gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopOutputStreamAnalyticsIntegrationTest.java [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+      out.write(expected);
+    }
+
+    assertThat(gcsFsIHelper.readFile(path)).isEqualTo(expected);
+  }
+
+  @Test
+  public void write_whenClosed_throwsIOException() throws Exception {
+    URI path = gcsFsIHelper.getUniqueObjectUri(getClass(), "write_after_closed");
+    GoogleHadoopFileSystem ghfs = setupGhfs("write_after_closed");
```
[the remark this answers is not in the document]
Comment (dheerajsngh) [reply gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopOutputStreamAnalyticsIntegrationTest.java]: Done.

File: gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopOutputStreamAnalyticsIntegrationTest.java [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+      out.write(initialContent);
+    }
+
+    assertThrows(IOException.class, () -> fs.create(hadoopPath, false));
+  }
+
+  @Test
+  public void write_directOutputStream_writesContentCorrectly() throws Exception {
+    URI path = gcsFsIHelper.getUniqueObjectUri(getClass(), "write_directOutputStream");
+    GoogleHadoopFileSystem ghfs = setupGhfs("write_directOutputStream");
```
[the remark this answers is not in the document]
Comment (dheerajsngh) [reply gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopOutputStreamAnalyticsIntegrationTest.java]: Done.

File: gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopOutputStreamAnalyticsIntegrationTest.java [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+      out.write(new byte[0], 0, 0);
+    }
+
+    assertThat(gcsFsIHelper.readFile(path)).isEqualTo(new byte[0]);
+  }
+
+  private GoogleHadoopFileSystem setupGhfs(String tag) throws Exception {
+    URI path = gcsFsIHelper.getUniqueObjectUri(getClass(), tag);
+    return GoogleHadoopFileSystemIntegrationHelper.createGhfs(path, getTestConfig());
+  }
```
[the remark this answers is not in the document]
Comment (dheerajsngh) [reply gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopOutputStreamAnalyticsIntegrationTest.java]: Done.

File: gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopOutputStreamAnalyticsIntegrationTest.java [hunk 5]
```diff
@@ -0,0 +1,237 @@
...
+
+  @Test
+  public void write_whenClosed_throwsIOException() throws Exception {
+    URI path = gcsFsIHelper.getUniqueObjectUri(getClass(), "write_after_closed");
+    GoogleHadoopFileSystem ghfs = setupGhfs("write_after_closed");
+
+    GoogleHadoopOutputStream out = createGhfsOutputStream(ghfs, path, CreateFileOptions.DEFAULT);
+    out.close();
+
+    assertThrows(IOException.class, () -> out.write(1));
```
[the remark this answers is not in the document]
Comment (dheerajsngh) [reply gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopOutputStreamAnalyticsIntegrationTest.java]: Done.

File: gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopOutputStreamAnalyticsIntegrationTest.java [hunk 6]
```diff
@@ -0,0 +1,237 @@
...
+    URI path = gcsFsIHelper.getUniqueObjectUri(getClass(), "overwrite_false");
+    Path hadoopPath = new Path(path);
+    FileSystem fs = GoogleHadoopFileSystemIntegrationHelper.createGhfs(path, getTestConfig());
+
+    byte[] initialContent = "initial content".getBytes(UTF_8);
+    try (FSDataOutputStream out = fs.create(hadoopPath, false)) {
+      out.write(initialContent);
+    }
+
+    assertThrows(IOException.class, () -> fs.create(hadoopPath, false));
```
[the remark this answers is not in the document]
Comment (dheerajsngh) [reply gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopOutputStreamAnalyticsIntegrationTest.java]: Done.

File: gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopOutputStreamAnalyticsIntegrationTest.java [hunk 7]
```diff
@@ -0,0 +1,236 @@
...
+
+  @Test
+  public void write_byteArray_writesContentCorrectly() throws Exception {
+    URI path = gcsFsIHelper.getUniqueObjectUri(getClass(), "write_byteArray");
+    Path hadoopPath = new Path(path);
+    FileSystem fs = GoogleHadoopFileSystemIntegrationHelper.createGhfs(path, getTestConfig());
+    byte[] expected = "hello analytics core write array".getBytes(UTF_8);
+
+    try (FSDataOutputStream out = fs.create(hadoopPath)) {
+      out.write(expected, 0, expected.length);
```
Comment (shrutisinghania) [inline gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopOutputStreamAnalyticsIntegrationTest.java:107]: We should add a test case that writes with a non-zero offset and a smaller length (e.g., out.write(expected, 5, 10)) to ensure the underlying Analytics Core write path properly respects the array slice and doesn't just read from the beginning of the buffer.

File: gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopOutputStreamAnalyticsIntegrationTest.java [hunk 8]
```diff
@@ -0,0 +1,236 @@
...
+    try (GoogleHadoopOutputStream out =
+        createGhfsOutputStream(ghfs, path, CreateFileOptions.DEFAULT)) {
+      out.write(new byte[0], 0, 0);
+    }
+
+    assertThat(gcsFsIHelper.readFile(path)).isEqualTo(new byte[0]);
+  }
+
+  private GoogleHadoopFileSystem setupGhfs(URI path) throws Exception {
+    return GoogleHadoopFileSystemIntegrationHelper.createGhfs(path, getTestConfig());
```
Comment (shrutisinghania) [inline gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopOutputStreamAnalyticsIntegrationTest.java]: Several test cases are still calling `GoogleHadoopFileSystemIntegrationHelper.createGhfs(path, getTestConfig())` directly. 
We can replace those calls with `setupGhfs(path)` for consistency across all test methods.

**Review by shrutisinghania — CHANGES_REQUESTED** [review]

File: gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopOutputStreamAnalyticsIntegrationTest.java (hunk 7 above)
Comment (dheerajsngh) [reply gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopOutputStreamAnalyticsIntegrationTest.java:107]: Added a test. Thanks

File: gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopOutputStreamAnalyticsIntegrationTest.java (hunk 8 above)
Comment (dheerajsngh) [reply gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopOutputStreamAnalyticsIntegrationTest.java]: Fixed. Thanks

File: gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopOutputStreamAnalyticsIntegrationTest.java [hunk 9]
```diff
@@ -0,0 +1,254 @@
...
+  @AfterClass
+  public static void afterClass() {
+    if (gcsFsIHelper != null) {
+      gcsFsIHelper.afterAllTests();
+    }
+  }
+
+  private Configuration getTestConfig() {
+    Configuration conf = GoogleHadoopFileSystemIntegrationHelper.getTestConfig();
+    conf.setBoolean("fs.gs.analytics.core.enable", true);
```
Comment (shrutisinghania) [inline gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopOutputStreamAnalyticsIntegrationTest.java]: nit:  Use the static property definitions in `GoogleHadoopFileSystemConfiguration` to prevent typos and ease future refactoring like
```conf.setBoolean(GCS_ANALYTICS_CORE_ENABLE.getKey(), true);```

**Review by shrutisinghania — APPROVED** [review]
As discussed, in a follow up PR the existing integration tests would be parameterized to run with analytics core.

File: gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopOutputStreamAnalyticsIntegrationTest.java [hunk 10]
```diff
@@ -0,0 +1,262 @@
...
+  public void hsync_writesContentCorrectly() throws Exception {
+    URI path = gcsFsIHelper.getUniqueObjectUri(getClass(), "hsync");
+    Path hadoopPath = new Path(path);
+    FileSystem fs = setupGhfs(path);
+    byte[] expected = new byte[5];
+    new Random().nextBytes(expected);
+
+    try (FSDataOutputStream fout = fs.create(hadoopPath)) {
+      for (int i = 0; i < expected.length; i++) {
+        fout.write(expected[i]);
```
Comment (animesh-g) [inline gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopOutputStreamAnalyticsIntegrationTest.java]: This can be flaky. GCS enforces 1 write per second on the same object name by default. It might result in 429. https://screenshot-v2.corp.google.com/7p943pi292eqg

File: gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopOutputStreamAnalyticsIntegrationTest.java [hunk 11]
```diff
@@ -0,0 +1,262 @@
...
+    assertThat(gcsFsIHelper.readFile(path)).isEqualTo(expected);
+  }
+
+  @Test
+  public void write_withZeroBufferSize_writesContentCorrectly() throws Exception {
+    URI path = gcsFsIHelper.getUniqueObjectUri(getClass(), "write_withZeroBufferSize");
+    Path hadoopPath = new Path(path);
+    Configuration config = getTestConfig();
+    config.setInt(GCS_OUTPUT_STREAM_BUFFER_SIZE.getKey(), 0);
+    FileSystem fs = GoogleHadoopFileSystemIntegrationHelper.createGhfs(path, config);
```
Comment (animesh-g) [inline gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopOutputStreamAnalyticsIntegrationTest.java]: Wrap fs in a try-with-resources block or manage a shared ghfs instance cleaned up in @After / @AfterClass. Applies to other tests as well

File: gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopOutputStreamAnalyticsIntegrationTest.java [hunk 12]
```diff
@@ -0,0 +1,262 @@
...
+    try (GoogleHadoopOutputStream out =
+        createGhfsOutputStream(ghfs, path, CreateFileOptions.DEFAULT)) {
+      out.write(expected);
+    }
+
+    assertThat(gcsFsIHelper.readFile(path)).isEqualTo(expected);
+  }
+
+  @Test
+  public void write_whenClosed_throwsClosedChannelException() throws Exception {
```
Comment (animesh-g) [inline gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopOutputStreamAnalyticsIntegrationTest.java]: I think this can be covered by unit test and integration test doesn't add any value here. This stream management happens at connector level and GCS doesn't play any role here. Check other tests as well for such redundancies.

File: gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopOutputStreamAnalyticsIntegrationTest.java [hunk 13]
```diff
@@ -0,0 +1,262 @@
...
+import org.junit.AfterClass;
+import org.junit.BeforeClass;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+/**
+ * Integration tests for GoogleHadoopOutputStream with Analytics Core write path enabled.
+ *
+ * <p>TODO: In a follow up PR the existing integration tests would be parameterized to run with
```
Comment (animesh-g) [inline gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopOutputStreamAnalyticsIntegrationTest.java]: I couldn't understand this TODO. If we want to parameterize the existing tests, why do we need a new test suite?

**Review by animesh-g — CHANGES_REQUESTED** [review]

File: gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopOutputStreamAnalyticsIntegrationTest.java (hunk 9 above)
Comment (dheerajsngh) [reply gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopOutputStreamAnalyticsIntegrationTest.java]: Done.

File: gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopOutputStreamAnalyticsIntegrationTest.java (hunk 10 above)
Comment (dheerajsngh) [reply gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopOutputStreamAnalyticsIntegrationTest.java]: Thanks for pointing it out. 
Update the test to call hsync only twice and used `fs.gs.outputstream.sync.min.interval` to put 1 sec interval between the write calls.

File: gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopOutputStreamAnalyticsIntegrationTest.java (hunk 11 above)
Comment (dheerajsngh) [reply gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopOutputStreamAnalyticsIntegrationTest.java]: Done.

File: gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopOutputStreamAnalyticsIntegrationTest.java (hunk 12 above)
Comment (dheerajsngh) [reply gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopOutputStreamAnalyticsIntegrationTest.java]: Makes sense. Removed the test from here.  Its already covered in  unit test.

File: gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopOutputStreamAnalyticsIntegrationTest.java (hunk 13 above)
Comment (dheerajsngh) [reply gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopOutputStreamAnalyticsIntegrationTest.java]: The intent of the TODO was to analyze in follow-up task if we can consolidate the tests for the default and the new analytics-core path and parametrize the tests to reduce code duplication. Updated the TODO to indicate the same.

**Review by animesh-g — APPROVED** [review]
